_sources/config.txt
環境設定
=================================
概要
--------------------
設定は `Lua <http://www.lua.org>`_ による設定ファイルに記載します。Windowsの場合iceberg実行ファイルと同一ディレクトリ、Linuxの場合 ``~/.iceberg`` の ``config.lua`` が読み込まれます。ヒストリファイルなどは設定ファイルと同一のディレクトリに保存されます。

また、Linuxの場合 ``~/.iceberg/config.d/*.lua`` ファイルが起動時に自動的に読み込まれるためこちらに設定を記載することも可能です。

設定ファイルにはいくつかのグローバル変数とグローバル関数を記載します。

キー名称
--------------------
この後の章で示す設定には以下のキー名称が使用できます。また、修飾キーと通常キーを ``-`` でつなぎ合わせることで同時押しを表現します。同時押しのキーは3つまでです。また、ホットキーの設定のみOSネイティブの仮想キーコードを受け付けます。

例：

    .. code-block:: lua

        "alt-space"
        "shift-ctrl-a"
        "0x1d" -- Windowsにおける「無変換」キー
        "0x1c" -- Windowsにおける「変換」キー

- 修飾キー
    - ``shift``
    - ``caps_lock``
    - ``ctrl``
    - ``alt``
    - ``num_lock``
    - ``meta``
    - ``scroll_lock``

- 通常キー
    - ``a`` ～ ``z``
    - ``0`` ～ ``9``
    - ``space``
    - ``backspace``
    - ``tab``
    - ``iso_key``
    - ``enter``
    - ``pause``
    - ``scroll_lock``
    - ``escape``
    - ``home``
    - ``left``
    - ``up``
    - ``right``
    - ``down``
    - ``page_up``
    - ``page_down``
    - ``end``
    - ``print``
    - ``insert``
    - ``menu``
    - ``help``
    - ``num_lock``
    - ``kp0`` : kpから始まるキーはテンキー上のキーです。
    - ``kp1``
    - ``kp2``
    - ``kp3``
    - ``kp4``
    - ``kp5``
    - ``kp6``
    - ``kp7``
    - ``kp8``
    - ``kp9``
    - ``kp_enter``
    - ``kp_last``
    - ``f1``
    - ``f2``
    - ``f3``
    - ``f4``
    - ``f5``
    - ``f6``
    - ``f7``
    - ``f8``
    - ``f9``
    - ``f10``
    - ``f11``
    - ``f12``
    - ``f_last``
    - ``shift_l``
    - ``shift_r``
    - ``control_l``
    - ``control_r``
    - ``caps_lock``
    - ``meta_l``
    - ``meta_r``
    - ``alt_l``
    - ``alt_r``
    - ``delete``
    - ``volume_down``
    - ``volume_mute``
    - ``volume_up``
    - ``media_play``
    - ``media_stop``
    - ``media_prev``
    - ``media_next``
    - ``home_page``
    - ``mail``
    - ``search``
    - ``back``
    - ``forward``
    - ``stop``
    - ``refresh``
    - ``sleep``
    - ``favorites``

systemグローバル変数
---------------------
設定例
~~~~~~~~~~~~~~~~~~~~~
以下に設定例と値の意味を示します。

    .. code-block:: lua

        system = {
          -- search pathから検索する際の深さのデフォルト値 --
          default_search_path_depth = 2,

          -- 補完候補のアイコンを表示するか(true:する, false:しない) --
          enable_icons = true,

          -- アイコンテーマ名:Linuxでのみ有効 --
          icon_theme = "nuoveXT.2.2",

          -- 補完候補のアイコンのキャッシュ数 -- 
          max_cached_icons = 9999,

          -- 何msキー入力が無い場合に補完候補を表示するか --
            -- 低速マシンの場合、自身のキータイプ間隔以上の値にすることにより
            -- 不要な補完動作を抑制することができます。
          key_event_threshold = 0,

          -- ヒストリの保存数 -- 
          max_histories = 500,

          -- 補完候補の表示数 -- 
          max_candidates = 15,

          -- クリップボード履歴の保存数:Windowsでのみ有効 -- 
          max_clipboard_histories = 15,

          -- 補完候補ソート時のヒストリの影響度合い(0.0~1.0)
          history_factor = 0.8,

          -- ディレクトリを開く際のコマンド、${1}にディレクトリパスが入る --
          file_browser = [[explorer ${1}]],

          -- ターミナルでコマンドを実行する際のコマンド、${1}コマンドが入る。ログインターミナルとして起動すること。:Linuxでのみ有効 --
          terminal = [[lxterminal -l -e ${1}]],

          -- 外部からコマンドを受け付けるポート(0: 無効)
          server_port = 13505,

          -- パス補完時に自動補完を有効にする
          path_autocomplete = true,

          -- 引数の補完時に自動補完を有効にする
          option_autocomplete = true,

          -- DirectWriteを無効にする:Windowsでのみ有効 --
          disable_direct_write = false,

          -- コマンド・ヒストリの補完時は自動補完は常に有効です --
        
          -- キー設定 --
          -- hot_keyはOS仮想キーコードでも可、単一キーも可。 --
          --   例: hot_key = "0x1d"  無変換 --
          hot_key = "ctrl-space",
          escape_key = "escape",
          list_next_key = "ctrl-n",
          list_prev_key = "ctrl-p",
          toggle_mode_key = "ctrl-r",
          kill_word_key = "ctrl-w",
        
          -- サーチパス --
          search_path = {
            {category="system", path = [[C:\Windows\System32]], depth = 1, pattern="^.*\\.(exe)$"}, 
            {category="programs", path = [[C:\Users\]] .. os.getenv("USERNAME") .. [[\AppData\Roaming\Microsoft\Windows\Start Menu\Programs]], depth = 10, pattern=[[^.*\.(exe|lnk)$]]},
            {category="programs", path = [[C:\ProgramData\Microsoft\Windows\Start Menu\Programs]], depth = 10, pattern=[[^.*\.(exe|lnk)$]]},
          },

          -- 補完設定 -- 
          completer = {
            -- コマンドの補完:あいまい一致 --
            command = ibs.COMP_ABBR,

            -- パスの補完:前方一致 -- 
            path    = ibs.COMP_BEGINSWITH,

            -- ヒストリの補完:部分一致 -- 
            history = ibs.COMP_PARTIAL,

            -- 引数の補完 -- 
            option  = ibs.COMP_PARTIAL,
        
            -- 引数補完のための関数 --
            option_func = {
              [":scan_search_path"] = function(values, pos)
                local candidates = {"all"}
                local keys       = {all = true}
                for i, value in ipairs(system.search_path) do
                  if value.category ~= nil and keys[value.category] == nil then
                    table.insert(candidates, value.category)
                    keys[value.category] = true
                  end
                end
                return candidates
              end
            }
          }
        }

サーチパス
~~~~~~~~~~~~~~~~~
サーチパスは、指定したディレクトリ配下を検索し自動的にコマンドとして登録する機能です。サーチパスの構成要素は以下です。

:category:
    サーチパスはカテゴリを持つことができます。指定したカテゴリのサーチパスのみを更新することが可能です。無指定の場合自動的に ``default`` というカテゴリに属します。

:path:
    検索する起点となるディレクトリです。

:depth:
    ``path`` から何階層検索するかを示す数値です。無指定の場合 ``system.default_search_path_depth`` が適応されます。

:pattern:
    コマンドとして登録するファイル名の正規表現パターン(完全一致)です。

補完関数
~~~~~~~~~~~~~~~~~
icebergではコマンドが入力された際の引数を補完する関数を定義できます。補完関数は以下のシグネチャです。

    .. code-block:: lua
        
        function(values, pos)
          return {"a", "b", "c"}
        end

        -- もしくは

        function(values, pos)
          return { 
           {value="a", icon="path_to/icon.png", description="desc"}, 
           {value="b", icon="path_to/icon.jpg", description="desc"},
           {value="c", icon="path_to/icon.gif", description="desc"}
          }
        end

``values`` は入力されている引数の配列です。引数が空の場合、空文字1要素になります。 ``pos`` は現在カーソルがある引数の配列における位置です。関数は補完候補を文字列のリストもしくは次の要素を含むテーブルのリストとして返す必要があります。文字列とテーブルを混在させることはできません。

:value:
    補完文字列です。この項目は必須です。
:icon:
    アイコンとして使用するファイルのパスです。
:description:
    説明として利用する文字列です。
:always_match:
    ``true`` を設定するとどんな入力にもマッチするようになり、選択時に入力欄が更新されなくなります。例えばWEB検索結果をただ表示する場合などに利用します。

補完関数は補完以外に情報表示だけのために利用することもできます。デフォルトの ``weather`` コマンドを参照してください。

またオプションを含む複雑な補完関数を書く場合は :lua:func:`icebergsupport.comp_state` も参照してください。


commandsグローバル変数
-----------------------
設定例
~~~~~~~~~~~~~~~~~~~~~
以下に代表的なコマンドの定義例を示します。

    .. code-block:: lua


        commands = { 
          -- ディレクトリ,ヒストリに残さない --
          windir = {path = [[C:\Windows]], history = false},

          -- 実行ファイル, icebergのカレントディレクトリで実行 -- 
          np = {path = [[notepad.exe]], description="Notepad", workdir="."},

          -- シェルスクリプト, ターミナルで実行する --
          np = {path = [[myscript.sh]], description="My script", terminal="yes"},

          -- lua関数, 補完関数あり -- 
          lua_sample = { 
            path = function(args) 
              local explorer = wins.foreground_explorer()
              if explorer then
                ibs.message(ibs.table_to_string(explorer))
              end
            end,
            completion = function(values, pos)
              return {"1","2","3"}
            end
            description="Sample Lua command"},


          -- URL, アイコン画像を指定 -- 
          google = { path = [[http://www.google.com/search?ie=utf8&q=${1}]], description=[[Searches words on Google]], history=false,
               icon = script_path ..[[images\google256.png]]},

          -- グループコマンド：連続してコマンドを実行 -- 
          group_sample = { path = ibs.group_command({"windir", {}}, {"np", {}}), description = "runs a group of commands"},
        
        }

コマンド
~~~~~~~~~~~~~~~~~~~~~~~
コマンドは以下の要素から構成されます。

:name: ``commands`` tableのキーとして表現されます。
:path: 
    実行対象のパスもしくはLuaの関数です。
    パスの場合、以下のように引数を含めることができます。また空白を含む場合は ``"`` で囲う必要があります。::

        path = [["C:\s p a c e\bin.exe" arg1 arg2]]

    また入力された変数を参照することができます。 ``google iceberg`` と入力された場合、 ``google`` コマンドの ``path`` が以下の場合、 ``${1}`` に ``iceberg`` が代入されます。::

        path = [[http://www.google.com/search?ie=utf8&q=${1}]]

    関数の場合、引数には文字列のリストが与えられます。関数は実行に成功した場合0を、失敗した場合は非0を返す必要があります。
:completion:
    ``system.completer.option_func`` と同じ形式の補完関数です。補完関数はコマンドでも ``system.completer.option_func`` でも登録できます。両方登録した場合はコマンドで定義したものが優先されます。
:description:
    補完候補ウインドウに表示される説明文です。
:icon:
    補完候補ウインドウに表示されるアイコン画像のパスです。Linuxの場合、アイコンテーマにおけるアイコン名でもかまいません。アイコン名の場合自動的に最適な画像を選択します。
:terminal:
    ターミナルで実行するかを示します。yesの場合常にターミナルで実行します。noの場合ターミナルで実行しません。autoの場合は自動的にターミナルで実行するか判定します。ターミナル実行コマンドは ``system.terminal`` で指定されたコマンドです。
:history:
    ``false`` を指定するとヒストリに残らなくなります。
:workdir:
    コマンドを実行するディレクトリです。以下の指定が可能です。

    - 固定値: そのディレクトリで実行されます。
    - ``.`` : icebergのカレントディレクトリで実行されます。
    - Lua関数: 関数の戻り値(文字列)のディレクトリで実行されます。例えば、外部ファイラのディレクトリを返す関数を設定すれば外部ファイラと連携できます。

shortcutsグローバル変数
------------------------
設定例
~~~~~~~~~~~~~~~~~~~~~
以下に代表的なショートカットの定義例を示します。

    .. code-block:: lua

        shortcuts = {
          { key = "ctrl-d", name = ":opendir" },
          { key = "ctrl-l", name = ":cd" }
        }

上記のようにショートカットを定義し、 ``c:\`` と入力欄に入力された状態で ``ctrl-l`` を押下したとします。その時以下のようにコマンドが実行されます。::

    :cd c:\

つまり、コマンドの引数として入力欄に入力されている値が渡されます。

on_key_upイベントハンドラ
--------------------------
キーが離された時に呼び出されます。

    .. code-block:: lua

        function on_key_up()
          local accept = 0
          return accept
        end

デフォルトの動作を抑止したい場合はこの関数で1を返してください。

on_key_downイベントハンドラ
---------------------------------
キーが押された時に呼び出されます。

    .. code-block:: lua

        function on_key_down()
          local accept = 0
          return accept
        end

デフォルトの動作を抑止したい場合はこの関数で1を返してください。

on_enterイベントハンドラ
--------------------------
enterキーが押下された際に呼び出されます。

    .. code-block:: lua

        function on_enter()
          local accept = 0
          return accept
        end

デフォルトの動作を抑止したい場合はこの関数で1を返してください。

on_initializeイベントハンドラ
--------------------------------
起動時に呼び出されます。

    .. code-block:: lua

        function on_enter()
          local error = 0
          return error
        end

この関数が1を返した場合、起動を停止します。


_sources/changes.txt
変更履歴
=======================
0.9.10 (2016-08-30)
-----------------------
- IMPROVED: 試験的にWindowsでDirectWriteを実装。パフォーマンス上の問題があるためデフォルトはGDIのままとする。

0.9.9からのアップグレード方法
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- 0.9.10をダウンロードしてください。
- 0.9.10から ``iceberg.exe`` をコピーして既存の0.9.9側に上書きしてください。
- ``config.lua`` の ``system`` グローバル変数に ``disable_direct_write=true``  を追加してください。DirectDraw機能を利用する場合は ``false`` をセットしてください。(ただし、現段階では少々描画が遅いです)

0.9.9 (2016-01-31)
-----------------------
- IMPROVED: 補完関数の第二引数に現在の引数の位置を渡すようになった。
- NEW: ``icebergsupport.table_find`` , ``icebergsupport.getopts`` , ``icebergsupport.comp_state`` を追加
- FIXED: アイコンロード時にデッドロックする問題を修正
- FIXED: キー入力時に稀にクラッシュする問題を修正

0.9.8からのアップグレード方法
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- 0.9.9をダウンロードしてください。
- 0.9.9から ``iceberg.exe`` をコピーして既存の0.9.8側に上書きしてください。
- 0.9.9から ``luamodule/icebergsupport.lua`` をコピーして既存の0.9.8側に上書きしてください。

0.9.8 (2015-12-27)
-----------------------
- NEW: コマンド定義に ``terminal`` 属性を追加。
- IMPROVED: Windowsにおいてシステム環境変数が変更された場合、icebergを再起動無しに以後起動したコマンドに変更された環境変数が反映されるようになった。
- IMPROVED: Pluginシステムを標準化した。
- IMPROVED: Windowsでも ``config.d`` ディレクトリ配下を読むようになった。
- FIXED: GIFアイコンとXPMアイコンを表示した時にクラッシュする問題を修正。

0.9.7からのアップグレード方法
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- 0.9.8をダウンロードしてください。
- 0.9.8から ``iceberg.exe`` をコピーして既存の0.9.7側に上書きしてください。
- 0.9.8から ``luamodule/icebergsupport.lua`` をコピーして既存の0.9.7側に上書きしてください。
- ``commands.cache`` と  ``icons.cache`` を削除してください。
- icebergディレクトリに ``plugins`` ディレクトリを作成してください。
- Windowsの場合、icebergディレクトリに ``config.d`` ディレクトリを作成してください。
- Windowsの場合、 ``config.lua`` の ``system`` グローバル変数に以下のように ``terminal``  を追加してください。

    .. code-block:: lua

        system = {
          (他の設定)
          terminal = "cmd /k ${1}",
          (他の設定)
        }

- ``config.lua`` の末尾に以下を追加してください。Linuxの場合、 末尾の ``load_config_d()`` は削除した上で追加してください(関数定義も削除して問題ありません)。

    .. code-block:: lua

        ibs.load_lua_files(ibs.join_path(script_path, "config.d"))
        ibs.load_plugins()


0.9.7 (2015-11-19)
-----------------------
- NEW: Linux上での動作をサポートした。
- NEW: アイコン画像にsvgを利用できるようになった。
- CHANGED: fltkのバージョンを1.3.3にアップグレード。
- CHANGED: 鬼車のバージョンを5.9.6にアップグレード。

0.9.6からのアップグレード方法
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- 0.9.7をダウンロードしてください。
- 0.9.7から ``iceberg.exe`` をコピーして既存の0.9.6側に上書きしてください。
- ``commands.cache`` と  ``icons.cache`` を削除してください。

0.9.6 (2014-10-20)
-----------------------
- FIXED: 一部のパスでアイコンが正しく読み込まれない問題を修正
- NEW: パス補完とオプション補完でオートコンプリートを有効にする ``path_autocomplete`` と ``option_autocomplete`` を追加
- FIXED: 一部のパスを読み込んだ際にクラッシュする問題を修正
- CHANGED: 使用するコンパイラをMinGW-W64 4.9.1にアップグレード
- IMPROVED: 単一キーをホットキーに割り当てられるようになった

0.9.5からのアップグレード方法
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- 0.9.6をダウンロードしてください。
- 0.9.6から ``iceberg.exe`` をコピーして既存の0.9.5側に上書きしてください。
- アイコン周りの修正が入っているので ``icons.cache`` を削除することをお勧めします。
- ``config.lua`` の ``system`` グローバル変数に以下のように ``path_autocomplete`` と ``option_autocomplete`` を追加してください。

    .. code-block:: lua

        system = {
          (他の設定)
          -- パス補完時にオートコンプリートを行うか
          path_autocomplete = true,
          -- オプション補完時にオートコンプリートを行うか
          option_autocomplete = true,
          (他の設定)
        }


0.9.5 (2014-03-04)
-----------------------
- FIXED: 補完候補ウインドウに ``'&'`` が描画できなかった問題を修正
- FIXED: 一部のキー組み合わせで動作しなかった問題を修正( ``shift-tab`` など )

0.9.4 (2013-11-05)
-----------------------
- NEW: 補完候補の属性に ``always_match`` を追加
- NEW: ``icebergsupport.selected_index`` , ``icebergsupport.brshift`` , ``icebergsupport.blshift`` を追加

0.9.3 (2013-11-01)
-----------------------
- FIXED: shift+矢印キーが動作しない問題を修正
- FIEED: 特定のアイテム上でコンテキストメニューを開こうとした時に落ちる問題を修正
- FIXED: 相対パスを絶対パスに変換する処理のバグを修正
- FIXED: ``server_port`` オプションを有効にしている状態で再起動に失敗する問題を修正
- CHANGED: コマンドの設定で ``history = false`` としてる場合もコマンド名のみのヒストリは取得するように変更
- IMPROVED: モーダルウインドウの扱いを改善(Enterキー押下でクローズできるようになった)

0.9.2 (2013-09-07)
-----------------------
- FIXED: スレッドセーフではない関数がマルチスレッド下で保護されていない問題を修正
- IMPROVED: コマンド定義でも補完関数を定義できるようになった
- NEW: ``icebergsupport.get_clipboard_histories`` を追加
    - 関連して ``system.max_clipboard_histories`` 設定値を追加
- NEW: ``clipboard`` コマンドをデフォルトコマンドに追加
- NEW: ``icebergsupport.add_history`` を追加
- NEW: ``-m activate`` 起動オプションを追加

0.9.1 (2013-08-24)
-----------------------
- FIXED: ショートカット実行時、入力欄に ``"`` で囲われた空白を含む値が合った場合の問題を修正
- FIXED: 補完関数がテーブルを返した場合、補完リストが入力に従いフィルタリングされなかった問題を修正
- IMPROVED: 相対パスの扱いを改善
- IMPROVED: 補完関数でjpegファイルをアイコンに使用できるようにした
- IMPROVED: アイコンキャッシュをより効率的な実装にした
- NEW: ``icebergsupport.unquote_path`` を追加
- NEW: ``alttab`` コマンドをデフォルトコマンドに追加

0.9.0 (2013-08-15)
-----------------------
- 公開開始


assets/windows/config.lua
local script_path = _iceberg_config_dir()
package.path = script_path .. [[\luamodule\?.lua;]] .. package.path
package.cpath = script_path ..[[\luamodule\?.dll;]] .. package.path
local ibs = require("icebergsupport")
local wins = require("winsupport")
local winalttab = require("winalttab")

-- configurations --
system = {
  default_search_path_depth = 2,
  enable_icons = true,
  max_cached_icons = 999999,
  key_event_threshold = 0,
  max_histories = 500,
  max_candidates = 15,
  max_clipboard_histories = 15,
  history_factor = 0.8,
  file_browser = [[explorer ${1}]],
  terminal  = [[cmd /k ${1}]],
  server_port = 40000,
  path_autocomplete = true,
  option_autocomplete = true,
  disable_direct_write = true,
  direct_write_params = "",

  hot_key = "ctrl-space",
  escape_key = "escape",
  list_next_key = "ctrl-n",
  list_prev_key = "ctrl-p",
  toggle_mode_key = "ctrl-r",
  kill_word_key = "ctrl-w",

  search_path = {
    {category="shortcuts", path = script_path .. [[\shortcuts]], depth = 5, pattern=[[^.*\.(exe|lnk|bat)$]]},
    {category="system", path = [[C:\Windows\System32]], depth = 1, pattern="^.*\\.(exe)$"}, 
    {category="programs", path = [[C:\Users\]] .. os.getenv("USERNAME") .. [[\AppData\Roaming\Microsoft\Windows\Start Menu\Programs]], depth = 10, pattern=[[^.*\.(exe|lnk)$]]},
    {category="programs", path = [[C:\ProgramData\Microsoft\Windows\Start Menu\Programs]], depth = 10, pattern=[[^.*\.(exe|lnk)$]]},
  },
  completer = {
    command = ibs.COMP_ABBR,
    path    = ibs.COMP_PARTIAL,
    history = ibs.COMP_PARTIAL,
    option  = ibs.COMP_PARTIAL,

    option_func = {}
  }
}

styles = {
  window_boxtype = ibs.BOXTYPE_BORDER_BOX, 
  window_posx  = nil,
  window_posy  = nil,
  window_width = 400,
  window_height = 50,
  window_padx  = 7,
  window_pady  = 7,
  window_bg_color = {238, 238, 238},
  window_alpha = 220,
  taskbar_height = 40,

  input_boxtype = ibs.BOXTYPE_FLAT_BOX,
  input_font  = "Verdana",
  input_font_size  = 24,
  input_font_color  = {238, 238, 238},
  input_bg_color  = {33, 33, 33},
  input_selection_bg_color = {99,99,99},

  list_boxtype = ibs.BOXTYPE_BORDER_BOX,
  list_padx = 5,
  list_pady = 5,
  list_border_color = {238,238,238},
  list_font  = "Verdana",
  list_font_size  = 15,
  list_desc_font_size  = 11,
  list_font_color  = {238, 238, 238},
  list_desc_font_color  = {220, 220, 220},
  list_selection_font_color  = {238, 238, 238},
  list_selection_desc_font_color  = {220, 220, 220},
  list_bg_color1  = {50, 50, 50},
  list_selection_bg_color1  = {99,99,99},
  list_bg_color2  = {33, 33, 33},
  list_selection_bg_color2  = {99,99,99},
  list_alpha = 220
}

commands = { 
  windir = {path = [[C:\Windows]]},
  np = {path = [[notepad.exe]], description="Notepad"},
  userdir = {path = [[C:\Users\]] .. os.getenv("USERNAME")},
  lua_sample = { path = function(args) 
    local explorer = wins.foreground_explorer()
    if explorer then
      ibs.message(ibs.table_to_string(explorer))
    end
  end ,description="Sample Lua command"},
  group_sample = { path = ibs.group_command({"userdir", {}}, {"np", {}}), description = "runs a group of commands"},


  -- default commands --
  [":version"] = {path = function(args) ibs.message(ibs.version()) end, description="show iceberg version", history=false},
  [":exit"] = {path = function(args) ibs.exit_application(0) end, description="exit iceberg", history=false},
  [":reboot"] = {path = function(args) ibs.reboot_application() end, description="reboot iceberg", history=false},
  [":pwd"] = {path = function(args) ibs.set_result_text(ibs.get_cwd()) end, description="show current work directory", history=false},
  [":cd"] = { path = function(args)
      if #args == 0 then
        ibs.message("Usage: :cd PATH")
        return 1
      end
      local path = ibs.to_directory_path(args[1])
      assert(ibs.set_cwd(path))
      ibs.set_result_text(ibs.get_cwd())
      return 0
    end, history = false},
  [":scan_search_path"] = {
    path = function(args) 
      if #args == 0 then
        ibs.message("Usage: :scan_search_path CATEGORY")
        return 1
      end
      ibs.scan_search_path(args[1]) 
    end, 
    completion = function(values, pos)
          local candidates = {"all"}
          local keys       = {all = true}
          for i, value in ipairs(system.search_path) do
            if value.category ~= nil and keys[value.category] == nil then
              table.insert(candidates, value.category)
              keys[value.category] = true
            end
          end
          return candidates
    end, 
    description="scan search paths to find commands", history=false},
  [":empty"] = {path = function(args) ibs.shell_execute(wins.cmd_path, {}, wins.foreground_explorer_path()) end, description = "open command prompt here", history=false, workdir = wins.foreground_explorer_path},
  [":opendir"] = { path = function(args)
      if #args == 0 then
        ibs.message("Usage: :opendir (command name | path)")
        return 1
      end
      local path = ibs.to_directory_path(args[1])
      assert(ibs.open_dir(path))
      return 0
    end, history = false},
  control_panel = {path = [[C:\WINDOWS\system32\control.exe]], description="Control panel"},
  windows_service = {path = [[C:\WINDOWS\system32\services.msc]], description=[[Service]]},
  my_computer = {path = [[::{20D04FE0-3AEA-1069-A2D8-08002B30309D}]], description=[[My computer]]},
  recycle_bin = {path = [[::{645FF040-5081-101B-9F08-00AA002F954E}]], description=[[Recycle bin]]},
  network_computer = {path = [[::{208D2C60-3AEA-1069-A2D7-08002B30309D}]], description=[[Network computer]]},
  google = { path = [[http://www.google.com/search?ie=utf8&q=${1}]], description=[[Searches words on Google]], history=false,
               icon = script_path ..[[\images\google256.png]]},
  mkdir = { path = function(args) 
      if #args == 0 then
        ibs.message("Usage: mkdir DIRNAME")
        return 1
      end
      local dirname = args[1]
      local fexplorer = wins.foreground_explorer()
      if not fexplorer then return 0 end
      local ok, out, err = ibs.command_output([[C:\Windows\System32\cmd.exe /c md "]] .. fexplorer.path.."\\" .. dirname .. [["]])
      if not ok then
        ibs.message(ibs.crlf2lf(ibs.local2utf8(err)))
        return 1
      end
      return 0
    end, description = "Usage: mkdir DIRNAME", history=false },
  cal = { path = function(args)
      local script = "ret = (" .. table.concat(args, " ") .. ")"
      local func = loadstring(script)
      if not func then
        ibs.message("Invalid expression.")
        return 1
      end
      func()
      ibs.set_result_text(ret)
      return 0
    end, description = "inline calculator", history = false},
  alttab = { 
    path = function(args) 
      if #args == 0 then return end
      local ok, r = ibs.regex_search([[\s\(ID:([0-9]+)\)]], Regex.NONE, args[1])
      if ok then
        winalttab.activate(tonumber(r:_1()))
      end
      return 0
    end, 
    completion = function(values, pos)
        local candidates = {}
        for i, tb in ipairs(winalttab.list()) do
          local title = tb.title .. " (ID:" .. tb.hwnd .. ")"
          local value = {value = title, description = tb.path, icon=tb.path}
          table.insert(candidates, value)
        end
        return candidates
    end,
    description = "move between open windows",
    history=false
  },
  clipboard = {
    path = function(args)
      if #args == 0 then return end
      local ok, r = ibs.regex_match([[\(([0-9]+)\).*]], Regex.NONE, args[1])
      if ok then
        ibs.set_clipboard(ibs.get_clipboard_histories()[tonumber(r:_1())])
      end
      return 0
    end,
    completion = function(values, pos)
        local candidates = {}
        for i, text in ipairs(ibs.get_clipboard_histories()) do
          local value = ibs.regex_gsub("\n", Regex.NONE, text, " ")
          value = ibs.regex_gsub("\t", Regex.NONE, value, "    ")
          local ok, r = ibs.regex_match("(.{50})(.*)", Regex.NONE, value)
          if ok then
            value = r:_1() .. " ... "
          end
          value = "(" .. tostring(i) .. ") : " .. value
          table.insert(candidates, value)
        end
        return candidates
    end,
    description = "clipboard history",
    history = false
  }
}

shortcuts = {
  { key = "ctrl-d", name = ":opendir" },
  { key = "ctrl-l", name = ":cd" }
}

function on_key_up()
  local accept = 0
  return accept
end

function on_key_down()
  local accept = 0
  if accept == 0 then
    accept = ibs.process_shortcut_keys() 
  end

  return accept
end

function on_enter()
  local accept = 0
  return accept
end

function on_initialize()
  local error = 0
  return error
end

-- load configurations from ./config.d/*.lua --
ibs.load_lua_files(ibs.join_path(script_path, "config.d"))
ibs.load_plugins()
